### PR TITLE
Optimize gcc

### PR DESCRIPTION
**Pull request by daschuer**
This reduces the "optimize" scons option too three and integrates the tuning option.
optimize: 
Set to:
  1 portable-binary: sse2 CPU (>= Pentium 4)
  2 native-binary: optimized exclusive for the CPU of this system
  3 legacy-binary: pure i386 code

Tested on Linux only 

https://bugs.launchpad.net/mixxx/+bug/1405678

### CODE REVIEW

**Comment by daschuer** [comment]
IMHO this is mergeable. Any additional comments?

**Comment by rryan** [comment]
@daschuer -- I haven't re-read your changes after our discussion about strings but if it still changes the integers then it is not mergeable for the reasons I mentioned already.

File: build/features.py [hunk 1]
```diff
@@ -1000,175 +992,81 @@ def configure(self, build, conf):
...
             build.env.Append(
                 CCFLAGS='-O3 -ffast-math -funroll-loops')
             if not int(build.flags['profiling']):
                 build.env.Append(CCFLAGS='-fomit-frame-pointer')
 
-            if optimize_level == 1:
-                # only includes what we already applied
-                self.status = "Enabled -- Basic Optimizations (-03)"
+            # Historicaly our gcc release packages are built with optimize=9.
+            if optimize_level == 1 or optimize_level == 9:
```
Comment (daschuer) [inline build/features.py]: Here is the legacy optimize level 9

**Comment by daschuer** [comment]
The conclusion was, that changing the levels to strings is to complicated for now.
I have re-added level 9 for gcc builds to be compatible.
Anything else todo?

**Comment by rryan** [comment]
> The conclusion was, that changing the levels to strings is to complicated for now.
> I have re-added level 9 for gcc builds to be compatible.
> Anything else todo?

Sorry to be strict but that means it's not ready to merge. Groups outside the project depend on those options.

**Comment by daschuer** [comment]
What can I do? 
Currently there is an exception if you select an uncommon option form the past, 1 and 9 is still working with a reasonable results, all other options which are most likely not used outside can be adapted in the user script.
raise Exception("optimize=%s is not supported, use 1 .. 3." % optimize_level)

**Comment by rryan** [comment]
What went wrong when you switched to semantic options instead of numbers?

```
optimize=off
optimize=portable
optimize=tuned
optimize=legacy
```

**Comment by daschuer** [comment]
Do not know exactly. See @kain88-de last comment, he has tried as well.

**Comment by rryan** [comment]
The optimize flag is only used in a couple of places -- the SoundTouch dependence, the Optimize feature, and the Tuned feature (and you've removed Tuned) -- so it shouldn't be hard.

**Comment by daschuer** [comment]
The rest of our build scripts is not aware of string properties, that is the hard part.

**Comment by kain88-de** [comment]
> The optimize flag is only used in a couple of places -- the SoundTouch dependence, the Optimize feature, and the Tuned feature (and you've removed Tuned) -- so it shouldn't be hard.

Yeah I thought so to. The problem is that we don't use the scons API for the CLI flags directly but build some sort of wrapper around (I guess by someone who adores C and never learned more of python then the syntax). Because if this we rely on the fact that the flags are numbered in various places of our build system. A proper change would also include changing that any valid besides `None` is valid. Currently only positive integers are valid.

You can have a look into the commit I linked above. I currently don't have time to look into this.

**Comment by rryan** [comment]
You don't have to use ```util.get_flags``` (I agree that function is jank). You can access ```SCons.Script.ARGUMENTS``` directly or add a ```get_string_flag``` method if you'd like.

**Comment by daschuer** [comment]
OK but I still do not know what you want exactly. On one hand you write.

> How about words for the new options? That way we can keep the sane option name and leave the old number-based cruft in for backwards compatibility -- or just make all numbers map to "portable".

On the other hand: 

> Sorry to be strict but that means it's not ready to merge. Groups outside the project depend on those options.

Since I do not have a special preference for one of these options, would you mind to adopt this branch and fix it for your needs? Thank you.

**Comment by ywwg** [comment]
I think he's saying the old number system should still work for legacy build configurations, but the new words should also work.

**Comment by kain88-de** [comment]
> You don't have to use util.get_flags (I agree that function is jank). You can access SCons.Script.ARGUMENTS directly or add a get_string_flag method if you'd like.

https://github.com/mixxxdj/mixxx/blob/59b3b467de26015bc0a5f6e1c82349a0e05c4860/src/SConscript.env#L123

That line is the problem. I'm not sure what `build_flags` is used for besides building the `build.h` file. Why are we even using that file. Would it mess with the build if we allow Strings there?

**Comment by ywwg** [comment]
build.h is included in version.cpp and is used to print out what build flags were used during compilation, for debugging, so it's not necessary for it to be programmatically perfect -- it just has to be readable.

my build.h contains the line 

#define BUILD_FLAGS "autodjcrates bulk hid localecompare mad optimize=9 qdebug shoutcast test vamp vinylcontrol"


so my guess is it's ok if it were to have optimize=somestring.  

All that list comprehension is doing is saying "go through the list of flags.  For every item whose value is greater than zero, print just the item if the value == 1, or print key=value if the value is greater than 1"

So what it's saying is for boolean flags, just print the name of the key.  For the optimize flag, print key=value.  It should be fine to rewrite this so that it still just prints the key name if the value is 1, or prints key=value if the value is > 1 or if it's not an int.  Don't try to do it in one line please :)

**Comment by rryan** [comment]
Right -- that's only used for printing the scons line when you boot mixxx. There's no harm in changing it.

**Comment by daschuer** [comment]
Thanks to @kain88-de, it is working now with strings.

* portable (alias 9): sse2 CPU (>= Pentium 4)
* native: optimized exclusive for the CPU of this system
* legacy: pure i386 code
* disabled: no optimization, for debug only

Now msvc debug builds are optimized for 'portable' as well. Since for me it makes no sense to test Mixxx in an unoptimized build.  Unoptimized code is useful to singe stepping, while looking for real had problems only. the rest should be happy with optimization. 

@rryan: Ready for merge?

**Comment by daschuer** [comment]
Added an option for building without any C flags 
build=none

**Comment by daschuer** [comment]
We have actually issue with the old flags 
https://bugs.launchpad.net/mixxx/+bug/1417582

@rryan: Is this merge-able now?

File: build/mixxx.py [hunk 2]
```diff
@@ -39,8 +39,8 @@ def __init__(self, target, machine, build, toolchain, available_features=[]):
         if build is None:
             build = 'debug'
 
-        if not build in ['debug', 'release']:
-            raise Exception("invalid build type")
+        if not build in ['debug', 'release', 'none']:
```
Comment (rryan) [inline build/mixxx.py]: Noooooooooo please don't! A build is either release or debug.

What version of Qt do you link against in a none build? 
What macro do you define (MIXXX_BUILD_DEBUG or MIXXX_BUILD_RELEASE) -- that affects the runtime assertions. 
What MSVC runtime do you link against?

I think the right way to do this is via a separate scons option.

File: build/features.py [hunk 3]
```diff
@@ -80,8 +80,9 @@ def configure(self, build, conf):
             conf.CheckLib(['rt', 'librt'])
 
             # -pthread tells GCC to do the right thing regardless of system
-            build.env.Append(CCFLAGS='-pthread')
-            build.env.Append(LINKFLAGS='-pthread')
+            if build.build_is_debug or build.build_is_release:
```
Comment (rryan) [inline build/features.py]: without -pthread our builds will crash

File: build/depends.py [hunk 4]
```diff
@@ -1001,12 +1002,14 @@ def configure(self, build, conf):
...
 
         if build.toolchain_is_gnu:
-            # Default GNU Options
-            build.env.Append(CCFLAGS='-pipe')
-            build.env.Append(CCFLAGS='-Wall')
-            build.env.Append(CCFLAGS='-Wextra')
-            # TODO(XXX) always generate debugging info?
-            build.env.Append(CCFLAGS='-g')
+            if build.build_is_debug or build.build_is_release:
+                build.env.Append(CPPDEFINES='MIXXX_BUILD_RELEASE')
```
Comment (rryan) [inline build/depends.py]: This drops all debug assertions in debug builds.

File: build/depends.py [hunk 5]
```diff
@@ -381,10 +381,11 @@ class SoundTouch(Dependence):
     SOUNDTOUCH_PATH = 'soundtouch-1.8.0'
 
     def sse_enabled(self, build):
-        optimize = int(util.get_flags(build.env, 'optimize', 1))
+        optimize = SCons.ARGUMENTS.get('optimize', 'portable')
         return (build.machine_is_64bit or
-                (build.toolchain_is_msvs and optimize > 2) or
-                (build.toolchain_is_gnu and optimize > 1))
+                (optimize == 'portable') or
+                (optimize == '9') or
```
Comment (rryan) [inline build/depends.py]: options 2 through 8 support SSE

File: build/features.py [hunk 6]
```diff
@@ -942,29 +943,34 @@ def description(self):
...
             return
 
-        optimize_level = int(build.flags['optimize'])
+        optimize_level = SCons.ARGUMENTS.get('optimize', 'portable')
+
+        if optimize_level == 'disabled':
+            self.status = "disabled: no optimazion, for debug only"
+            return
+
+        if not build.build_is_debug and not build.build_is_release and optimize_level == '':
```
Comment (rryan) [inline build/features.py]: Why not provide an explicit optimize=none option?

Also, this requires the user to say ```scons optimize=``` because the default is portable so not providing optimize means portable. May as well make it an explicit option with optimize=none.

Comment (rryan) [inline build/features.py]: Derp, you already have a 'disabled' option. Ok -- so I think this block doesn't really need to be here.

File: build/features.py [hunk 7]
```diff
@@ -994,180 +1000,85 @@ def configure(self, build, conf):
...
-            elif optimize_level >= 1:
-                self.status = "Enabled -- Maximize Speed (/O2)"
-                build.env.Append(CCFLAGS='/O2')
-            # elif optimize_level >= 2:
-            #    self.status = "Enabled -- Maximum Optimizations (/Ox)"
-            #    build.env.Append(CCFLAGS = '/Ox')
+            build.env.Append(CCFLAGS='/O2')
+
+            # Historicaly our release packages are built with optimize=9.
+            if optimize_level == 'portable' or optimize_level == '' or optimize_level == 9:
```
Comment (rryan) [inline build/features.py]: ```optimize_level == 9``` will never be true since optimize_level is a string.

File: build/features.py [hunk 8]
```diff
@@ -994,180 +1000,85 @@ def configure(self, build, conf):
...
+                self.status = "portable: sse2 CPU (>= Pentium 4)"
+                build.env.Append(CCFLAGS='/arch:SSE2')
+                build.env.Append(CPPDEFINES=['__SSE__', '__SSE2__'])
+            elif optimize_level == 'native':
+                self.status = "native: exclusive for this CPU (%s)" % build.machine
+                build.env.Append(CCFLAGS='/favor:' + build.machine)
+            elif optimize_level == 'legacy':
+                self.status = "legacy: pure i386 code"
+            else:
+                raise Exception("optimize={} is not supported. "
```
Comment (rryan) [inline build/features.py]: This will throw exceptions if someone uses the numbers. Could you please check for the legacy numbers above and map them to something sensible?

File: build/features.py [hunk 9]
```diff
@@ -994,180 +1000,85 @@ def configure(self, build, conf):
...
                 build.env.Append(
-                    CPPDEFINES=['__SSE__', '__SSE2__', '__SSE3__'])
-            elif optimize_level == 9:
-                self.status = "Enabled (Tuned Generic)"
-                # This option is for release builds packaged for 64-bit. We
-                # don't know what kind of CPU they'll have, so let
-                # -mtune=generic pick the most common, but compatible options. 
+                    CCFLAGS='-mtune=generic')
+                # -mtune=generic pick the most common, but compatible options.
                 # Used by the debian rules script.
```
Comment (rryan) [inline build/features.py]: Don't think this is true -- (our rules script uses portable).

File: build/mixxx.py (hunk 2 above)
Comment (daschuer) [inline build/mixxx.py]: This was a request from Nico, I was requesting a mode without any compiler flags. 
What happens and if it is useful or not, is up to the caller. 
It should emulate the cMake feature: 
CMAKE_BUILD_TYPE = None   (CMAKE_CXX_FLAGS or CMAKE_C_FLAGS used) 
I think it should be a release mode by default.

File: build/features.py (hunk 3 above)
Comment (daschuer) [inline build/features.py]: This may happen if on is building with "none". But maybe he wants that.

File: build/depends.py (hunk 5 above)
Comment (daschuer) [inline build/depends.py]: Now they are gone. If one has a build based on the old options, he has decide new, that is intended.

File: build/mixxx.py (hunk 2 above)
Comment (rryan) [inline build/mixxx.py]: Yes, I realize why the change was made. I'm saying I don't think it's the right way to accomplish the desire of a no-flags build. The code is more complicated if we have to consider build 3 build types. There's no "none" MSVC runtime -- the user needs to pick release or debug.

File: build/features.py (hunk 6 above)
Comment (daschuer) [inline build/features.py]: If one checks out mixxx and build with just 
scons 
he should get an SSE optimized debug build

This if branch is executed when the user has explicit forced no flags by build=none.
So I think I can remove the  < and optimize_level == ''" > part

File: build/features.py (hunk 7 above)
Comment (daschuer) [inline build/features.py]: Phyton does the magic for us. It just works.

File: build/features.py (hunk 8 above)
Comment (daschuer) [inline build/features.py]: This is intended.

**Comment by daschuer** [comment]
noted addressed.

File: build/depends.py (hunk 5 above)
Comment (rryan) [inline build/depends.py]: That's not good -- what I was saying is that it's not ok to break build scripts / etc. You can easily migrate the numbers to the new text options.

File: build/features.py (hunk 7 above)
Comment (rryan) [inline build/features.py]: Python doesn't do this.
```Python
>>> '9' == 9
False
```

File: build/mixxx.py (hunk 2 above)
Comment (daschuer) [inline build/mixxx.py]: I think the release version is picked in that case, since we always check only for the debug flag.
I do not bother much if a plain build=none gives not a sensible result, because that is not the use case for the option.
For me it is more the last resort to fix an exotic problem on a currently unsupported build platform without editing the scons scripts,  but passing magic flags on the command line.

File: build/features.py (hunk 7 above)
Comment (daschuer) [inline build/features.py]: Than probably scons, because it works. But I will add the '9' version.

File: build/mixxx.py (hunk 2 above)
Comment (rryan) [inline build/mixxx.py]: Right, but if the person who wrote that check didn't realize there were 3 options and did the opposite check then it would select the debug version. The point is it's entirely arbitrary and it does things the original programmer did not explicitly intend. Adding support for a flag-less build is fine but this way of doing it makes it easy to write bugs. 

It's especially concerning if this none option is what will be used by Linux repository maintainers so we need to be precise about what we are doing.

File: build/features.py (hunk 7 above)
Comment (rryan) [inline build/features.py]: SCons.ARGUMENTS contains a plain old Python str. Here's a debug trace from inside the Optimize feature:
```Python
value = SCons.ARGUMENTS.get('optimize', '')
import pdb
pdb.set_trace()
(Pdb) type(value)
<type 'str'>
(Pdb) value == '9'
True
(Pdb) value == 9
False
```

If it works, something else is making it work :).

**Comment by daschuer** [comment]
I do not have the overview of the demand of the different build setups. 
@rryan: Can you adopt this branch? Feel free to remove the build=none option and map the old numbers as it is required. Thank you.

File: build/features.py [hunk 10]
```diff
@@ -942,29 +943,34 @@ def description(self):
         return "Optimization and Tuning"
 
     def enabled(self, build):
-        build.flags['optimize'] = util.get_flags(build.env, 'optimize', 1)
-        if int(build.flags['optimize']):
+        build.flags['optimize'] = SCons.ARGUMENTS.get('optimize', '')
+        if build.flags['optimize'] is not 'portable':
```
Comment (rryan) [inline build/features.py]: ```is not``` is for referential equality, you want ```==``` here

Comment (rryan) [inline build/features.py]: Also, why is this enabled when optimize isn't equal to 'portable'? You want configure to run for all values of optimize so I think you should return True from this function or ```build != 'disabled'```

Comment (kain88-de) [inline build/features.py]: This would mean that we only disable the feature with the keyword 'disable'. There is no warning raised if someone uses '0' or 'off' or 'false'. The pythonic way to solve this is to define a set of allowed optimize values and then check if the given value is in the set

```Python
from sets import Set
self.optimize_options = Set(['portable ', 'tuned', '9', '8'])
build.flags['optimize'] = SCons.ARGUMENTS.get('optimize', None)
return build.flags['optimize'] in optimize_options
```

Then we will return if the given option is valid. Otherwise optimization is turned off.

Comment (daschuer) [inline build/features.py]: Is 
```Python
def enabled(self, build):
```
actually used? 

Since we have an assertion on unsupported values. We can just check for 
```optimize != 'disabled'```

Comment (rryan) [inline build/features.py]: Yep!
https://github.com/mixxxdj/mixxx/blob/master/src/SConscript.env#L80

**Comment by rryan** [comment]
> I do not have the overview of the demand of the different build setups. 
> @rryan: Can you adopt this branch? Feel free to remove the build=none option and map the old numbers as it is required. Thank you.

Sure -- sorry for the unclear requirements.

File: build/features.py (hunk 10 above)
Comment (kain88-de) [inline build/features.py]: something like 'off' shouldn't trigger an assertion. 'disable' is not intuitive for people coming from other build sytems
